# Getting and Knowing Data

In [12]:
# import library
import pandas as pd
import numpy as np

## Step 1: Import Dataset

In [13]:
df = pd.read_csv("chipotle.tsv", sep = "\t")

In [15]:
# xem thu 5 dong dau tien file no nhu the nao
df.head(5)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98

## Step 2: Dataset Overview
(cái nhìn tổng quát)

In [59]:
# coi có tất cả bao nhiu hàng, mỗi hàng có bao nhiu cột
df.shape

(4622, 5)

In [60]:
#xem thử kiểu dữ liệu nó như thế nào
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB


- Cột order_id có 4622 dòng, không có giá trị null, kiểu dữ liệu là integer
- Cột choice_description có 3376 non_null giá trị => có nhiều missing data, kiểu dữ liệu là object
- Bộ nhớ sử dụng là 180.7+ KB



In [61]:
#print all columns's name (in tên cột)
# df.columns đây là dạng object nên chuyển thành list cho dễ nhìn
list(df.columns)

['order_id', 'quantity', 'item_name', 'choice_description', 'item_price']

In [62]:
#xem in dex xem data có bị ngắt quảng không
df.index

RangeIndex(start=0, stop=4622, step=1)

In [17]:
# staticstic summary, chỉ return mean, std,... với kiểu dữ liệu integer 
# thêm inlude = "all" để thống kê tất cả các biến
=df.describe(include = "all")

order_id     quantity     item_name choice_description item_price
count   4622.000000  4622.000000          4622               3376       4622
unique          NaN          NaN            50               1043         78
top             NaN          NaN  Chicken Bowl        [Diet Coke]     $8.75 
freq            NaN          NaN           726                134        730
mean     927.254868     1.075725           NaN                NaN        NaN
std      528.890796     0.410186           NaN                NaN        NaN
min        1.000000     1.000000           NaN                NaN        NaN
25%      477.250000     1.000000           NaN                NaN        NaN
50%      926.000000     1.000000           NaN                NaN        NaN
75%     1393.000000     1.000000           NaN                NaN        NaN
max     1834.000000    15.000000           NaN                NaN        NaN

## Phương pháp cắt phổ biển loc và iloc
- loc = Location: Vị trí
- iloc = Index Location: Vị trí theo Index
- Giúp access những cái row và columns dựa theo điều kiện mà mình muốn specify

In [64]:
# Select all rows with multiple condictions: Chọn hàng với nhiều điều kiện khác nhau
# Tìm hàng mà có oder_quantity maximum(15 cái)
# loc nhận 2 input, nếu nhập 1 cái đầu tiên thì chỉ lọc coi hàng nào có giá trị quantity có giá trị lớn hơn 15
# df.loc[hàng,biến(cột) == điều kiện], ngoài ra có thể thêm nhiều điều kiện khác nữa
# mở ngoặc đóng ngoặc, sau dấu phẩy ngoặc vuông để specify
df.loc[(df.quantity >= 2) & (df.item_name == "Nantucket Nectar"), ["order_id", "quantity", "item_name"]]

order_id  quantity         item_name
601        247         2  Nantucket Nectar
2379       947         2  Nantucket Nectar

In [65]:
# Chọn hàng theo vị trí
# type(df.iloc[9]) đang ở dạng series nên thêm 1 [] để về dạng df
df.iloc[[9]]

order_id  quantity      item_name  \
9         5         1  Steak Burrito   

                                  choice_description item_price  
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25

In [66]:
df.iloc[3:11] #specify từ số 3 đến 10

order_id  quantity                              item_name  \
3          1         1  Chips and Tomatillo-Green Chili Salsa   
4          2         2                           Chicken Bowl   
5          3         1                           Chicken Bowl   
6          3         1                          Side of Chips   
7          4         1                          Steak Burrito   
8          4         1                       Steak Soft Tacos   
9          5         1                          Steak Burrito   
10         5         1                    Chips and Guacamole   

                                   choice_description item_price  
3                                                 NaN     $2.39   
4   [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
5   [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98   
6                                                 NaN     $1.69   
7   [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75   
8   [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...     $9.25   
9   [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25   
10                                                NaN     $4.45

In [67]:
# model machine learning model thid split thành x và y để pridict
# iloc[từ index đến index, cột cần lấy] 
x = df.iloc[3:5, :-1]
y = df.iloc[3:5, -1] #lấy cột cuối cùng
print(y)

3     $2.39 
4    $16.98 
Name: item_price, dtype: object


## Data Manipulation
Thao tác với dữ liệu

In [68]:
df.item_price.dtype

dtype('O')

## Hàm Apply


In [69]:
# thay doi $ thanh empty string
# chỉ thay đổi nhất thời, muốn thay đổi luôn thì gán ngược lại vào df
# lưu ngược lại cột là đc, easy lắm
df.item_price.apply(lambda x: float(x.replace('$', '')))

0        2.39
1        3.39
2        3.39
3        2.39
4       16.98
        ...  
4617    11.75
4618    11.75
4619    11.25
4620     8.75
4621     8.75
Name: item_price, Length: 4622, dtype: float64

In [70]:
df.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98

In [71]:
df.item_price = df.item_price.apply(lambda x: float(x.replace('$', '')))

In [72]:
df.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description  item_price  
0                                                NaN        2.39  
1                                       [Clementine]        3.39  
2                                            [Apple]        3.39  
3                                                NaN        2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98

## Tạo 1 hoặc nhiều columns mới dựa trên dữ liệu có sẵn


In [81]:
df["total_price"] = df["quantity"] * df["item_price"]
#df.total_price = df.quantity * df.item_price

In [82]:
df.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description  item_price  total_price  
0                                                NaN        2.39         2.39  
1                                       [Clementine]        3.39         3.39  
2                                            [Apple]        3.39         3.39  
3                                                NaN        2.39         2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98        33.96

## How much  was the revenue for the period in the dataset?
 - Tổng toàn bộ doanh thu trong cái dataset này là bao nhiêu

In [84]:
revenue = df.total_price.sum()
print(revenue)

39237.02


## Which was the most-order item
- món hàng nào được order nhiều nhất

## Group By

In [98]:
most_oder = df.groupby("item_name")["quantity"].sum()
most_oder.sort_values(ascending = False).head(5)

item_name
Chicken Bowl           761
Chicken Burrito        591
Chips and Guacamole    506
Steak Burrito          386
Canned Soft Drink      351
Name: quantity, dtype: int64

## Tìm Unique Value

In [19]:
#value_counts() đếm tất cả các món hàng
#count() tổng cộng là bao nhiu món hàng được bán ra
df.item_name.value_counts().count()

50

In [103]:
# nunique() các món hàng độc nhất
df.item_name.nunique()

50